tests/Microsoft.Bot.Builder.TestBot/Startup.cs

```diff
@@ -32,45 +33,52 @@ public Startup(IHostingEnvironment env)
-
-                var conversationState = new ConversationState(dataStore);
+            //services.AddSingleton<IAdapterIntegration>(sp =>
+            //{
+            //    var options = sp.GetRequiredService<IOptions<BotFrameworkOptions>>().Value;
+            //    var accessors = sp.GetRequiredService<TestBotAccessors>();
 
-                var accessors = new TestBotAccessors
-                {
-                    ConversationDialogState = conversationState.CreateProperty<DialogState>("DialogState"),
-                    ConversationState = new ConversationState(dataStore)
-                };
+            //    options.Middleware.Add(new AutoSaveStateMiddleware(accessors.ConversationState));
+            //    options.Middleware.Add(new ShowTypingMiddleware());
 
-                return accessors;
-            });
+            //    var botFrameworkAdapter = new BotFrameworkAdapter(options.CredentialProvider, options.ChannelProvider, options.ConnectorClientRetryPolicy, options.HttpClient)
+            //    {
+            //        OnTurnError = options.OnTurnError,
+            //    };
 
-            services.AddSingleton<IAdapterIntegration>(sp =>
-            {
-                var options = sp.GetRequiredService<IOptions<BotFrameworkOptions>>().Value;
-                var accessors = sp.GetRequiredService<TestBotAccessors>();
+            //    foreach (var middleware in options.Middleware)
+            //    {
+            //        botFrameworkAdapter.Use(middleware);
+            //    }
 
-                options.Middleware.Add(new AutoSaveStateMiddleware(accessors.ConversationState));
-                options.Middleware.Add(new ShowTypingMiddleware());
+            //    //return botFrameworkAdapter;
 
-                var botFrameworkAdapter = new BotFrameworkAdapter(options.CredentialProvider, options.ChannelProvider, options.ConnectorClientRetryPolicy, options.HttpClient)
-                {
-                    OnTurnError = options.OnTurnError,
-                };
-
-                foreach (var middleware in options.Middleware)
-                {
-                    botFrameworkAdapter.Use(middleware);
-                }
+            //    return new InteceptorAdapter(botFrameworkAdapter);
+            //});
 
-                //return botFrameworkAdapter;
+            IStorage dataStore = new MemoryStorage();
+            var conversationState = new ConversationState(dataStore);
 
-                return new InteceptorAdapter(botFrameworkAdapter);
-            });
+            var accessors = new TestBotAccessors
+            {
+                ConversationDialogState = conversationState.CreateProperty<DialogState>("DialogState"),
+                ConversationState = conversationState
+            };
 
-            services.AddBot<TestBot>();
+            services.AddBot<IBot>(
+                (IServiceProvider sp) =>
+                {
+                    return new TestBot(accessors);
```

Comment:
So, again... not really the focus of this PR right now, but now that you have this overload, you can just get rid of the extra `XXXAccessors` class pattern and hand in the exact dependencies you want explicitly:

```CSharp
var dataStore = new MemoryStorage(); 
var conversationState = new ConversationState(dataStore);
var dialogStateProperty = conversationState.CreateProperty<DialogState>(nameof(DialogState));

services.AddBot(
    sp => new TestBot(conversationState, dialogStateProperty),
   options =>
   {
      // … elided for brevity ...
   });
```

Reply:
Yes absolutely - TestBot is just test code - and I totally agree with the "accessors" pattern being clumsy. In fact I nearly made this change myself - but then realized I also squirreled a semaphore away on that object just to make debugging easier. (The whole semaphore thing is a whole other discussion.)

I'll add a unit test for this behavior.